tiledb/sm/tile/chunk_buffers.cc

```diff
@@ -0,0 +1,429 @@
+    nbytes_read += bytes_to_read;
+
+    chunk_offset = 0;
+    ++chunk_idx;
+  }
+
+  return Status::Ok();
+}
+
+Status ChunkBuffers::write(
+    const void* const buffer, const uint64_t nbytes, const uint64_t offset) {
+  if ((offset + nbytes) > size()) {
+    return Status::TileError("Chunk write error; write out of bounds");
+  }
+
+  size_t chunk_idx;
+  size_t chunk_offset;
+  RETURN_NOT_OK(translate_logical_offset(offset, &chunk_idx, &chunk_offset));
+
+  uint64_t nbytes_written = 0;
+  while (nbytes_written < nbytes) {
+    void* chunk_buffer = buffers_[chunk_idx];
+    if (!chunk_buffer) {
+      if (contigious_) {
+        return Status::TileError("Chunk write error; unset contigious buffer");
+      } else {
+        RETURN_NOT_OK(alloc_discrete(chunk_idx, &chunk_buffer));
+        buffers_[chunk_idx] = chunk_buffer;
+      }
+    }
+
+    const uint64_t nbytes_remaining = nbytes - nbytes_written;
+    const uint64_t cbytes_remaining = get_chunk_size(chunk_idx) - chunk_offset;
+    const uint64_t bytes_to_write =
+        std::min(nbytes_remaining, cbytes_remaining);
+
+    memcpy(
+        static_cast<char*>(chunk_buffer) + chunk_offset,
+        static_cast<const char*>(buffer) + nbytes_written,
+        bytes_to_write);
+    nbytes_written += bytes_to_write;
+
+    chunk_offset = 0;
+    ++chunk_idx;
+  }
+
+  return Status::Ok();
+}
+
+Status ChunkBuffers::write(const ChunkBuffers& rhs, const uint64_t offset) {
+  LOG_FATAL(
+      "ChunkBuffers::write(const ChunkBuffers &, uint64_t) unimplemented");
+
+  // TODO REMOVE: bypass compiler warning
+  if (offset || rhs.size()) {
+  }
+
+  return Status::Ok();
+}
+
+uint32_t ChunkBuffers::get_chunk_size(const size_t chunk_idx) const {
+  assert(chunk_idx < buffers_.size());
+  if (fixed_chunk_sizes()) {
+    return chunk_idx == (buffers_.size() - 1) ? last_chunk_size_ : chunk_size_;
+  } else {
+    return var_chunk_sizes_[chunk_idx];
+  }
+}
+
+bool ChunkBuffers::fixed_chunk_sizes() const {
+  return var_chunk_sizes_.empty();
+}
+
+Status ChunkBuffers::translate_logical_offset(
+    const uint64_t logical_offset,
+    size_t* const chunk_idx,
+    size_t* const chunk_offset) const {
+  assert(chunk_idx);
+  assert(chunk_offset);
+
+  // Optimize for the common case.
+  if (logical_offset == 0) {
+    *chunk_idx = 0;
+    *chunk_offset = 0;
+  }
+
+  if (fixed_chunk_sizes()) {
+    *chunk_idx = logical_offset / chunk_size_;
+    *chunk_offset = logical_offset % chunk_size_;
+  } else {
+    // The expectation is that the number of chunks is
+    // sufficiently small that we can perform an O(N)
+    // lookup. If the number of chunks is abnormally large,
+    // this could become a performance bottleneck. Let's
+    // assert() here that the number of chunks is less than
+    // 32k. That ensures we are operating on a 'var_chunk_sizes_'
+    // object that is roughly half the maximum size of a 256KB
+    // L1 cache.
+    assert(var_chunk_sizes_.size() < 32000);
```

Comment:
I don't think we need this assertion. I find the `O(N)` algorithm inevitable for the case of var-sized chunks anyway.